src/lustre/lustreDesugarIfBlocks.ml

```diff
@@ -0,0 +1,330 @@
+          | A.AnnotProperty (pos, _, _) 
+          | A.AnnotMain (pos, _) -> mk_error pos (MisplacedNodeItemError ni)
+        )
+      | A.IfBlock (pos, cond, [], ni::nis) -> (
+        match ni with
+          | A.Body (Equation (_, lhs, expr)) -> 
+            (* Update corresponding tree (or add new tree if it doesn't exist) *)
+            let trees = LhsMap.update lhs 
+              (fun tree -> match tree with
+                | Some tree -> Some (add_eq_to_tree (conds @ [(false, cond)]) expr tree)
+                | None -> Some (add_eq_to_tree (conds @ [(false, cond)]) expr (Leaf None))) 
+              trees 
+            in
+            (helper (A.IfBlock (pos, cond, [], nis)) trees conds)
+          (* Recurse, keeping track of our location within the if blocks using the 
+             'conds' list. *)
+          | A.IfBlock _ -> 
+            let* res = (helper ni trees (conds @ [(false, cond)])) in
+            (helper (A.IfBlock (pos, cond, [], nis))
+                   res
+                   conds)
+          (* Misplaced frame block, annot main, or annot property *)
+          | A.FrameBlock (pos, _, _, _)
+          | A.Body (Assert (pos, _)) 
+          | A.AnnotProperty (pos, _, _)
+          | A.AnnotMain (pos, _) -> mk_error pos (MisplacedNodeItemError ni)
+        )
+      (* We've processed everything in the if block. *)
+      | A. IfBlock (_, _, [], []) -> R.ok (trees)
+      (* shouldn't be possible *)
+      | _ -> assert false
+  ) in
+  (helper ib LhsMap.empty [])
+  
+(** Converts a tree of conditions/expressions to an ITE expression. *)
+let rec tree_to_ite pos node =
+  match node with
+    | Leaf Some expr -> expr
+    | Leaf None -> A.Ident(pos, HString.mk_hstring "ib_oracle")
+    | Node (left, cond, right) -> 
+      TernaryOp (pos, Ite, cond, tree_to_ite pos left, tree_to_ite pos right)
+
+(** Returns the type associated with a tree. *)
+let get_tree_type ctx lhs = 
+  match lhs with
+    | A.StructDef(_, [SingleIdent(_, i)]) -> (Ctx.lookup_ty ctx i) 
+    | A.StructDef(_, [ArrayDef(_, i, _)]) -> (
+      match (Ctx.lookup_ty ctx i) with
+        (* Assignment in the form of A[i] = f(i), so the RHS type is no
+           longer an array *)
+        | Some (ArrayType (_, (ty, _))) -> Some ty
+        | _ -> None
+      )
+    (* Other cases not possible *)
+    | _ -> assert false
+
+(** Fills empty spots in an ITE with oracles. *)
+let rec fill_ite_with_oracles ite ty = 
+  match ite with
+    | A.TernaryOp (pos, Ite, cond, e1, e2) -> 
+      let e1, gids1, decls1 = (
+      match e1 with
+        | TernaryOp _ as top -> fill_ite_with_oracles top ty
+        | Ident(p, s) when s = HString.mk_hstring "ib_oracle" -> 
+          let (expr, gids) = (mk_fresh_ib_oracle ty) in
+          (match expr with
+            | A.Ident (_, name) ->  expr, gids, [A.NodeVarDecl(p, (p, name, ty, ClockTrue))]
+            | _ -> assert false (* not possible*))
+        | _ -> e1, GI.empty (), []
+      ) in 
```

Comment:
Since the same code is applied to both branches, factor out this code into a local function, and call the function on each branch. It makes the symmetry clear, keeping the code synchronized is easier, etc.

Reply:
I chose a different refactoring that I think makes a bit more sense. Let me know if you want me to switch back to your suggestion.

Reply:
LGTM. Thanks.